File: src/airscope/ui/log_translate.py
"""Plain-English translator for technical log lines.

``plain_english(line)`` maps terse capture/attack output into
human-readable labels that non-expert users can understand at a glance.
The mapping is deterministic and stateless.
"""
from __future__ import annotations

# Each entry: (technical substring → plain-English replacement).
# Order matters: more-specific patterns come first so shorter
# prefixes don't shadow them.
_RULES: list[tuple[str, str]] = [
    # ── handshake / 4-way ──────────────────────────────────────────
    ("Got EAPOL M1", "Handshake step 1 captured"),
    ("Got EAPOL M2", "Handshake step 2 captured"),
    ("Got EAPOL M3", "Handshake step 3 captured"),
    ("Got EAPOL M4", "Handshake step 4 captured"),
    ("Crackable pair confirmed", "Complete handshake — ready to crack"),
    ("Handshake already saved", "Already have a capture for this network"),
    ("HANDSHAKE captured", "Handshake captured — ready to crack"),
    ("✓ HANDSHAKE", "Handshake captured — ready to crack"),
    # ── PMKID ──────────────────────────────────────────────────────
    ("PMKID found", "Key fingerprint captured silently"),
    ("PMKID captured", "Key fingerprint captured silently"),
    ("✓ PMKID", "Key fingerprint captured silently"),
    # ── WEP ────────────────────────────────────────────────────────
    ("WEP KEY captured", "WEP key captured"),
    ("✓ WEP KEY", "WEP key captured"),
    ("Decloaked Hidden Network", "Hidden network revealed"),
    # ── WPS ────────────────────────────────────────────────────────
    ("WPS PIN found", "WPS PIN found"),
    ("✓ WPS PIN", "WPS PIN found"),
    ("WPS PSK (via PushButton)", "Password recovered via WPS button"),
    ("WPS PSK found", "Password recovered via WPS"),
    ("✓ WPS PSK", "Password recovered via WPS"),
    ("WPS locked", "Router has locked WPS — too many attempts"),
    # ── SAE / WPA3 ────────────────────────────────────────────────
    ("SAE captured", "WPA3 login exchange captured"),
    ("✓ SAE", "WPA3 login exchange captured"),
    # ── EvilTwin ───────────────────────────────────────────────────
    ("PSK recovered", "Password found!"),
    ("Password found", "Password found!"),
    # ── deauth / disassoc ──────────────────────────────────────────
    ("Sending deauth to client", "Disconnecting a device from the network"),
    ("Deauth of", "Disconnecting devices from"),
    ("Deauth stopped", "Deauthentication stopped"),
    ("✓ Deauth provoked a crackable handshake",
     "Handshake captured after disconnecting devices"),
    # ── scanning / channel ─────────────────────────────────────────
    ("Channel hopping", "Scanning across all channels"),
    ("Tuned to channel", "Switched to channel"),
    ("Tried to tune to channel", "Attempted to switch to channel"),
    ("Passively listening", "Listening for traffic silently"),
    # ── target / focus ─────────────────────────────────────────────
    ("Target acquired", "Locked onto target network"),
    ("Encryption:", "Encryption:"),
    ("BSSID:", "BSSID:"),
    # ── PMF ────────────────────────────────────────────────────────
    ("PMF Required", "Network requires management frame protection"),
    ("Deauth attacks have been disabled",
     "Disconnect attacks blocked — network uses management protection"),
    # ── WEP attacks ────────────────────────────────────────────────
    ("ChopChop", "Packet forgery attack"),
    ("ARP Replay", "Replaying ARP packets to generate data"),
    # ── EvilTwin campaign ──────────────────────────────────────────
    ("EvilTwin of", "Fake network targeting"),
    ("EvilTwin stopped", "Fake network attack stopped"),
    ("twin live on ch", "Fake network running on channel"),
    # ── WPS attacks ────────────────────────────────────────────────
    ("WPS PIN brute started", "Trying common WPS PINs against the router"),
    ("WPS PushButton", "WPS button-press detection"),
    # ── SAE capture ────────────────────────────────────────────────
    ("SAE capture on", "Capturing WPA3 login from"),
    # ── batch ──────────────────────────────────────────────────────
    ("Batch already running", "An automated attack sequence is already running"),
    ("Nothing to attack", "No targetable networks in the selection"),
    ("Batch done", "Automated attack sequence finished"),
    # ── silence ────────────────────────────────────────────────────
    ("AP Silenced", "Network notifications paused"),
    ("AP UnSilenced", "Network notifications resumed"),
    # ── capture events ─────────────────────────────────────────────
    ("Existing captures in", "Previous captures found in"),
]


def plain_english(line: str) -> str:
    """Translate a technical log line to plain English.

    Scans ``_RULES`` for the first matching substring; returns the
    plain-English replacement. If no rule matches the original line
    is returned unchanged.
    """
    for technical, friendly in _RULES:
        if technical in line:
            return friendly
    return line

File: webui/src/lib/log_translate.ts
/**
 * Plain-English translator for technical log lines.
 *
 * Mirrors the Python ``airscope.ui.log_translate.plain_english()`` exactly.
 * Rules are applied in order; first match wins.
 */

interface Rule {
  technical: string;
  friendly: string;
}

const RULES: Rule[] = [
  // ── handshake / 4-way ──────────────────────────────────────────
  { technical: "Got EAPOL M1", friendly: "Handshake step 1 captured" },
  { technical: "Got EAPOL M2", friendly: "Handshake step 2 captured" },
  { technical: "Got EAPOL M3", friendly: "Handshake step 3 captured" },
  { technical: "Got EAPOL M4", friendly: "Handshake step 4 captured" },
  { technical: "Crackable pair confirmed", friendly: "Complete handshake — ready to crack" },
  { technical: "Handshake already saved", friendly: "Already have a capture for this network" },
  { technical: "HANDSHAKE captured", friendly: "Handshake captured — ready to crack" },
  { technical: "✓ HANDSHAKE", friendly: "Handshake captured — ready to crack" },
  // ── PMKID ──────────────────────────────────────────────────────
  { technical: "PMKID found", friendly: "Key fingerprint captured silently" },
  { technical: "PMKID captured", friendly: "Key fingerprint captured silently" },
  { technical: "✓ PMKID", friendly: "Key fingerprint captured silently" },
  // ── WEP ────────────────────────────────────────────────────────
  { technical: "WEP KEY captured", friendly: "WEP key captured" },
  { technical: "✓ WEP KEY", friendly: "WEP key captured" },
  { technical: "Decloaked Hidden Network", friendly: "Hidden network revealed" },
  // ── WPS ────────────────────────────────────────────────────────
  { technical: "WPS PIN found", friendly: "WPS PIN found" },
  { technical: "✓ WPS PIN", friendly: "WPS PIN found" },
  { technical: "WPS PSK (via PushButton)", friendly: "Password recovered via WPS button" },
  { technical: "WPS PSK found", friendly: "Password recovered via WPS" },
  { technical: "✓ WPS PSK", friendly: "Password recovered via WPS" },
  { technical: "WPS locked", friendly: "Router has locked WPS — too many attempts" },
  // ── SAE / WPA3 ────────────────────────────────────────────────
  { technical: "SAE captured", friendly: "WPA3 login exchange captured" },
  { technical: "✓ SAE", friendly: "WPA3 login exchange captured" },
  // ── EvilTwin ───────────────────────────────────────────────────
  { technical: "PSK recovered", friendly: "Password found!" },
  { technical: "Password found", friendly: "Password found!" },
  // ── deauth / disassoc ──────────────────────────────────────────
  { technical: "Sending deauth to client", friendly: "Disconnecting a device from the network" },
  { technical: "Deauth of", friendly: "Disconnecting devices from" },
  { technical: "Deauth stopped", friendly: "Deauthentication stopped" },
  { technical: "✓ Deauth provoked a crackable handshake", friendly: "Handshake captured after disconnecting devices" },
  // ── scanning / channel ─────────────────────────────────────────
  { technical: "Channel hopping", friendly: "Scanning across all channels" },
  { technical: "Tuned to channel", friendly: "Switched to channel" },
  { technical: "Tried to tune to channel", friendly: "Attempted to switch to channel" },
  { technical: "Passively listening", friendly: "Listening for traffic silently" },
  // ── target / focus ─────────────────────────────────────────────
  { technical: "Target acquired", friendly: "Locked onto target network" },
  { technical: "Encryption:", friendly: "Encryption:" },
  { technical: "BSSID:", friendly: "BSSID:" },
  // ── PMF ────────────────────────────────────────────────────────
  { technical: "PMF Required", friendly: "Network requires management frame protection" },
  { technical: "Deauth attacks have been disabled", friendly: "Disconnect attacks blocked — network uses management protection" },
  // ── WEP attacks ────────────────────────────────────────────────
  { technical: "ChopChop", friendly: "Packet forgery attack" },
  { technical: "ARP Replay", friendly: "Replaying ARP packets to generate data" },
  // ── EvilTwin campaign ──────────────────────────────────────────
  { technical: "EvilTwin of", friendly: "Fake network targeting" },
  { technical: "EvilTwin stopped", friendly: "Fake network attack stopped" },
  { technical: "twin live on ch", friendly: "Fake network running on channel" },
  // ── WPS attacks ────────────────────────────────────────────────
  { technical: "WPS PIN brute started", friendly: "Trying common WPS PINs against the router" },
  { technical: "WPS PushButton", friendly: "WPS button-press detection" },
  // ── SAE capture ────────────────────────────────────────────────
  { technical: "SAE capture on", friendly: "Capturing WPA3 login from" },
  // ── batch ──────────────────────────────────────────────────────
  { technical: "Batch already running", friendly: "An automated attack sequence is already running" },
  { technical: "Nothing to attack", friendly: "No targetable networks in the selection" },
  { technical: "Batch done", friendly: "Automated attack sequence finished" },
  // ── silence ────────────────────────────────────────────────────
  { technical: "AP Silenced", friendly: "Network notifications paused" },
  { technical: "AP UnSilenced", friendly: "Network notifications resumed" },
  // ── capture events ─────────────────────────────────────────────
  { technical: "Existing captures in", friendly: "Previous captures found in" },
];

/**
 * Translate a technical log line to plain English.
 * Returns the original line if no rule matches.
 */
export function plainEnglish(line: string): string {
  for (const rule of RULES) {
    if (line.includes(rule.technical)) {
      return rule.friendly;
    }
  }
  return line;
}
